netperf_tool/src/server_node.cpp

```diff
@@ -0,0 +1,225 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "server_node.hpp"
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/logging.hpp"
+
+#include "utils.hpp"
+
+namespace netperf_tool
+{
+
+ServerNode::ServerNode(
+  const rclcpp::NodeOptions & options,
+  const rclcpp::QoS & sub_qos)
+: Node("server", "netperf_tool", options)
+{
+  sub_ = this->create_subscription<netperf_tool_interfaces::msg::Bytes>(
+    "test_topic", sub_qos,
+    std::bind(&ServerNode::handle_msg, this, std::placeholders::_1, std::placeholders::_2));
+  srv_ = this->create_service<netperf_tool_interfaces::srv::GetResults>(
+    "get_results",
+    std::bind(
+      &ServerNode::handle_get_results_request,
+      this,
+      std::placeholders::_1,
+      std::placeholders::_2));
+}
+
+void
+ServerNode::handle_msg(
+  const rclcpp::SerializedMessage & serialized_msg,
```

Comment:
Why are you using the serialized version of the callback? To try and avoid that playing into the latency calculations?

Reply:
To be able to get the serialized size easily, without having to serialize again.

> To try and avoid that playing into the latency calculations?

That's also a benefit.